File: drive-java/src/main/java/frc/robot/commands/Performance.java
package frc.robot.commands;

import edu.wpi.first.wpilibj.command.Command;
import edu.wpi.first.wpilibj.smartdashboard.SmartDashboard;
import frc.math.Stats;

/**
* Used to measure the performance time it takes for the main loop to run.
*/
public class Performance extends Command {
  private Stats stats;
  private double lastTime;
  
  public Performance() {
    stats = new Stats();
    this.setRunWhenDisabled(true);
  }
  
  public void addMeasurement(double timeSecs) {
    lastTime = timeSecs;
    stats.add(timeSecs);
  }
  
  // Called just before this Command runs the first time
  @Override
  protected void initialize() {
    lastTime = 0.0;
    stats.zero();
  }
  
  // Called repeatedly when this Command is scheduled to run
  @Override
  protected void execute() {
    SmartDashboard.putNumber("Run Last", lastTime);
    SmartDashboard.putNumber("Run Avg", stats.getAvg());
    SmartDashboard.putNumber("Run Max", stats.getMax());
  }
  
  // Make this return true when this Command no longer needs to run execute()
  @Override
  protected boolean isFinished() {
    return false;
  }
  
}

File: drive-cpp/src/main/cpp/commands/Performance.cpp
#include "commands/Performance.h"
#include <frc/smartdashboard/SmartDashboard.h>
#include "Robot.h"

Performance::Performance() : m_lastSecs(0),
                             m_stats() {
  m_lastSecs = 0;
}

// Called just before this Command runs the first time
void Performance::Initialize() {
  m_stats.Zero();
  m_lastSecs = 0;
}

// Called repeatedly when this Command is scheduled to run
void Performance::Execute() {
  frc::SmartDashboard::PutNumber("Run Last", m_lastSecs);
  frc::SmartDashboard::PutNumber("Run Avg", m_stats.GetAvg());
  frc::SmartDashboard::PutNumber("Run Max", m_stats.GetMax());
}

// Make this return true when this Command no longer needs to run execute()
bool Performance::IsFinished() {
  return false;
}
